File: take_u_forward/graph/32_Shortest Path in Weighted undirected graph.java
import java.util.*;

class Solution {

    public List<Integer> shortestPath(int n, int m, int[][] edges) {

        // Adjacency list: node -> {adjNode, weight}
        List<List<int[]>> adj = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            adj.add(new ArrayList<>());
        }

        for (int[] e : edges) {
            adj.get(e[0]).add(new int[]{e[1], e[2]});
            adj.get(e[1]).add(new int[]{e[0], e[2]});
        }

        // Min heap {distance, node}
        PriorityQueue<long[]> pq =
                new PriorityQueue<>(Comparator.comparingLong(a -> a[0]));

        long[] dist = new long[n + 1];
        Arrays.fill(dist, Long.MAX_VALUE);

        int[] parent = new int[n + 1];
        for (int i = 1; i <= n; i++) parent[i] = i;

        dist[1] = 0;
        pq.add(new long[]{0, 1});

        while (!pq.isEmpty()) {
            long[] cur = pq.poll();
            long dis = cur[0];
            int node = (int) cur[1];

            // 🔥 IMPORTANT optimization
            if (dis > dist[node]) continue;

            for (int[] it : adj.get(node)) {
                int adjNode = it[0];
                int wt = it[1];

                if (dis + wt < dist[adjNode]) {
                    dist[adjNode] = dis + wt;
                    parent[adjNode] = node;
                    pq.add(new long[]{dist[adjNode], adjNode});
                }
            }
        }

        // No path
        if (dist[n] == Long.MAX_VALUE)
            return List.of(-1);

        // Reconstruct path
        List<Integer> path = new ArrayList<>();
        int node = n;
        while (parent[node] != node) {
            path.add(node);
            node = parent[node];
        }
        path.add(1);
        Collections.reverse(path);

        // Required format
        List<Integer> ans = new ArrayList<>();
        ans.add((int) dist[n]); // total weight
        ans.addAll(path);

        return ans;
    }
}

File: take_u_forward/graph/32_Shortest Path in Weighted undirected graph.cpp
class Solution {
public:
    vector<int> shortestPath(int n, int m, vector<vector<int>>& edges) {

        // Adjacency list
        vector<vector<pair<int,int>>> adj(n + 1);
        for (auto &e : edges) {
            adj[e[0]].push_back({e[1], e[2]});
            adj[e[1]].push_back({e[0], e[2]});
        }

        // Min heap {distance, node}
        priority_queue<pair<long long,int>, 
            vector<pair<long long,int>>, 
            greater<pair<long long,int>>> pq;

        vector<long long> dist(n + 1, 1e18);
        vector<int> parent(n + 1);

        for (int i = 1; i <= n; i++)
            parent[i] = i;

        dist[1] = 0;
        pq.push({0, 1});

        while (!pq.empty()) {
            auto [d, node] = pq.top();
            pq.pop();

            if (d > dist[node]) continue;

            for (auto &it : adj[node]) {
                int adjNode = it.first;
                int wt = it.second;

                if (d + wt < dist[adjNode]) {
                    dist[adjNode] = d + wt;
                    parent[adjNode] = node;
                    pq.push({dist[adjNode], adjNode});
                }
            }
        }

        // No path
        if (dist[n] == 1e18)
            return {-1};

        // Reconstruct path
        vector<int> path;
        int node = n;
        while (parent[node] != node) {
            path.push_back(node);
            node = parent[node];
        }
        path.push_back(1);
        reverse(path.begin(), path.end());

        // Required output format
        vector<int> ans;
        ans.push_back(dist[n]);      // total weight
        for (int x : path) ans.push_back(x);

        return ans;
    }
};
